Lh5443 fix uploading checksum time consuming issue

ChanYiLin opened the pull request:
[longhorn/longhorn#5443](https://github.com/longhorn/longhorn/issues/5443)

When after syncing file there is a post-processing function `finishProcessing` which will get file checksum and write config.
`GetFileChecksum` may be time consuming and can lead to client timeout if file is large.

This PR move it to a goroutine to make the api return to client first. After calculating the checksum, it will still update the state of syncfile to ready and then Longhorn will automatically do the following job (e.g. transfer file to manager)

ChanYiLin commented:
Test Result

Phase1
- As you can see from `POST` request, when the upload finished, server responded immediately (waiting: 465ms)
- The state was still in-progress
- Data source server was still calculating the checksum
![in-progress](https://user-images.githubusercontent.com/7100904/223676897-3c7c4e60-97ea-4185-8a59-da8abadbf305.png)


Phase2
- Data source server finished calculating checksum and saved config file
- State became ready
![ready](https://user-images.githubusercontent.com/7100904/223676979-8c3a8b12-45c8-4cef-970b-92edc027f6ad.png)

ChanYiLin commented on pkg/sync/sync_file.go:
```diff
@@ -656,37 +656,70 @@ func (sf *SyncingFile) finishProcessingNoLock(err error) (finalErr error) {
+		// if the state is already failed, there is no need to update the state
+		if sf.state == types.StateFailed {
+			return
+		}
```
unlock is called with defer here [#L677](https://github.com/longhorn/backing-image-manager/pull/105/files#diff-66b1ec3193d0ca8bd7fa4ee54caab93c7e0ed80edc090504487134b579bd2a71R677)

ChanYiLin replied:
sorry for the misleading, the commit you requested change is not the final version. I have refactored and fixed it in the next commit.

ChanYiLin commented on pkg/sync/sync_file.go:
```diff
@@ -656,37 +656,70 @@ func (sf *SyncingFile) finishProcessingNoLock(err error) (finalErr error) {
+			return
+		}
+
+		sf.writeConfigNoLock()
```
as above reply, the whole postProcessSyncFile is lock at the beginning and unlock in defer

ChanYiLin replied:
Sorry for the misleading, the commit you requested change is not the final version.
I have refactored and fixed it in the next commit.

ChanYiLin commented on pkg/sync/sync_file.go:
```diff
@@ -663,43 +663,44 @@ func (sf *SyncingFile) finishProcessingNoLock(err error) (finalErr error) {
-		sf.writeConfigNoLock()
+	sf.lock.Lock()
+	defer sf.lock.Unlock()
+	sf.currentChecksum, err = util.GetFileChecksum(sf.tmpFilePath)
```
Actually, original code has already put the whole finishProcessing(including GetFileChecksum) inside a lock
https://github.com/longhorn/backing-image-manager/blob/master/pkg/sync/sync_file.go#L609-L615
So this is an existing issue?

shuo-wu replied:
When the goroutine calculates the checksums, other threads for other requests cannot fetch this file info. That's what I am concerning about.

ChanYiLin replied:
So I can include this fix in this PR

shuo-wu replied:
To move the checksuming outside of the lock, it's better to compare the ModTime before and after the calculation. This help verifies if the file is modified during the calculation.

ChanYiLin commented:
Hi @shuo-wu I have moved the calculating checksum out of lock and squashed all commits into one to make it easy for review.
Thanks

Hi @innobead, I am curious about the codefactor standard. It shows the code is complex but I have no clue how to fix it, any suggestions, thanks!

innobead commented:
> Hi @innobead, I am curious about the codefactor standard. It shows the code is complex but I have no clue how to fix it, any suggestions, thanks!

@ChanYiLin What CodeFactor suggests does not necessarily mean we need to fix it that way, as it depends on different contexts. However, it can give us an idea of how to improve the code based on the existing style and pattern.

c3y1huang commented:
> > Hi @innobead, I am curious about the codefactor standard. It shows the code is complex but I have no clue how to fix it, any suggestions, thanks!
> 
> @ChanYiLin What CodeFactor suggests does not necessarily mean we need to fix it that way, as it depends on different contexts. However, it can give us an idea of how to improve the code based on the existing style and pattern.

You can also follow the CodeFactor suggestions to see what is reasonable to refactor and lower the complexity.

![Screenshot-2023-03-10-14:43:52](https://user-images.githubusercontent.com/22583541/224243607-c5db197a-ec39-4efa-a4b6-c0ec2c42444b.png)

In your case, maybe abstract file checksum validation logics in `postProcessSyncFile` to a separate method.

c3y1huang commented on pkg/sync/sync_file.go:
```diff
@@ -656,39 +647,75 @@ func (sf *SyncingFile) finishProcessingNoLock(err error) (finalErr error) {
+		return
+	}
+	if modificationTime := util.FileModificationTime(sf.tmpFilePath); modificationTime != sf.modificationTime {
+		finalErr = errors.Wrapf(err, "tmp file %v has been modified while calculaing checksum", sf.tmpFilePath)
```
Looks like err is nil, no need to wrap.

ChanYiLin replied:
fixed

c3y1huang reviewed (changes requested):
BTW, there are 2 signed-offs in the commit message. And also suggest to include the issue number.

shuo-wu commented on pkg/sync/sync_file.go:
```diff
@@ -656,39 +647,75 @@ func (sf *SyncingFile) finishProcessingNoLock(err error) (finalErr error) {
+		finalErr = errors.Wrapf(err, "failed to calculate checksum for tmp file %v", sf.tmpFilePath)
+		return
+	}
+	if modificationTime := util.FileModificationTime(sf.tmpFilePath); modificationTime != sf.modificationTime {
+		finalErr = fmt.Errorf("tmp file %v has been modified while calculaing checksum", sf.tmpFilePath)
+		return
+	}
```
This comparison should be in the lock

ChanYiLin replied:
fixed, thanks!

shuo-wu reviewed (commented):
Others LGTM

c3y1huang commented:
> feat(#5443): make getting file checksum async in post processing

Suggest avoiding prefixing the issue number with `#`, this could potentially create a hyperlink to the issue in the project. You can reference how other commit messages add the issue number.

Ref: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

ChanYiLin commented:
> > feat(#5443): make getting file checksum async in post processing
> 
> Suggest avoiding prefixing the issue number with `#`, this could potentially create a hyperlink to the issue in the project. You can reference how other commit messages add the issue number.
> 
> Ref: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

thanks for the advise, fixed

c3y1huang commented on pkg/sync/sync_file.go:
```diff
@@ -656,39 +647,75 @@ func (sf *SyncingFile) finishProcessingNoLock(err error) (finalErr error) {
 		return
 	}
 
+	sf.log.Infof("SyncingFile: succeeded processing file")
```
```suggestion
	sf.log.Info("SyncingFile: succeeded processing file")
```

ChanYiLin replied:
fixed, thanks

c3y1huang commented on pkg/sync/sync_file.go:
```diff
@@ -656,39 +647,75 @@ func (sf *SyncingFile) finishProcessingNoLock(err error) (finalErr error) {
 	}
 
+	sf.log.Infof("SyncingFile: succeeded processing file")
+	return
```
Redundant.

ChanYiLin replied:
fixed, thanks

c3y1huang reviewed (approved):
LGTM, Thanks!

innobead commented:
@Mergifyio backport v1.4.x v1.3.x
